packages/dynamic_layouts/lib/src/wrap_layout.dart

```diff
@@ -0,0 +1,270 @@
+class _RunMetrics {
+  _RunMetrics({
+    required this.maxSliver,
+    required this.currentSizeUsed,
+    required this.scrollOffset,
+  });
+
+  double maxSliver;
+  double currentSizeUsed;
+  double scrollOffset;
+}
+
+/// Controls the layout of tiles in a grid.
+/// A [SliverGridLayout] that uses dynamically sized tiles.
+///
+/// Rather that providing a grid with a [SliverGridLayout] directly, you instead
+/// provide the grid a [SliverGridDelegate], which can compute a
+/// [SliverGridLayout] given the current [SliverConstraints].
+///
+/// This layout is used by [SliverGridDelegateWithWrapping].
+///
+/// See also:
+///
+///  * [SliverGridDelegateWithWrapping], which uses this layout.
+///  * [SliverGridLayout], which represents an arbitrary tile layout.
+///  * [DynamicSliverGridGeometry], which represents the size and position of a
+///     single tile in a grid.
+///  * [SliverGridDelegate.getLayout], which returns this object to describe the
+///    delegate's layout.
+///  * RenderDynamicSliverGrid], which uses this class during its
+///    RenderDynamicSliverGrid.performLayout] method.
+class SliverGridWrappingTileLayout extends DynamicSliverGridLayout {
+  /// Creates a layout that uses dynamic sized and spaced tiles.
+  ///
+  /// All of the arguments must not be null and must not be negative.
+  SliverGridWrappingTileLayout({
+    required this.mainAxisSpacing,
+    required this.crossAxisSpacing,
+    required this.childMainAxisExtent,
+    required this.childCrossAxisExtent,
+    required this.crossAxisExtent,
+    required this.scrollDirection,
+  })  : assert(mainAxisSpacing != null && mainAxisSpacing >= 0),
+        assert(crossAxisSpacing != null && crossAxisSpacing >= 0),
+        assert(childMainAxisExtent != null && childMainAxisExtent >= 0),
+        assert(childCrossAxisExtent != null && childCrossAxisExtent >= 0),
+        assert(crossAxisExtent != null && crossAxisExtent >= 0),
+        assert(scrollDirection != null &&
+            (scrollDirection == Axis.horizontal ||
+                scrollDirection == Axis.vertical));
+
+  /// The direction in wich the layout should be built.
+  final Axis scrollDirection;
+
+  /// The extent of the child in the non-scrolling axis.
+  final double crossAxisExtent;
+
+  /// The number of logical pixels between each child along the main axis.
+  final double mainAxisSpacing;
+
+  /// The number of logical pixels between each child along the cross axis.
+  final double crossAxisSpacing;
+
+  /// The number of pixels from the leading edge of one tile to the trailing
+  /// edge of the same tile in the main axis.
+  final double childMainAxisExtent;
+
+  /// The number of pixels from the leading edge of one tile to the trailing
+  /// edge of the same tile in the cross axis.
+  final double childCrossAxisExtent;
+
+  /// The model that is used internally to keep track of how much space is left
+  /// and how much has been used.
+  final List<_RunMetrics> model = <_RunMetrics>[
```

Comment:
does this need to be publicly accessible? Or is this private to this file? In the latter case, make this private by starting the name with an underscore?

Reply:
Actually, since `_RunMetrics` is private, "model" must be private as well. There should be a lint for it. You may have to rebase your PR to the latest `main` branch to get those warnings.